### PR TITLE
Scene Collections V2: Backup and Sync

### PR DESCRIPTION
**Pull request by andycreeth**
A rewrite of the scene collections service to support multiple user accounts and backup/sync.

### CODE REVIEW

File: app/services/scene-collections/index.ts [hunk 1]
```diff
@@ -0,0 +1,887 @@
...
+    channel: number;
+  }[];
+}
+
+/**
+ * V2 of the scene collections service:
+ * - Completely asynchronous
+ * - Server side backup
+ */
+export class SceneCollectionsService extends StatefulService<ISceneCollectionsManifest> {
```
Comment (holiber) [inline app/services/scene-collections/index.ts]: We have to expose some API for Streamdeck. To do that create a file with `#serviceName#-api.ts` name and define an interface SceneCollectionsServiceApi. List the methods to expose and add some jsdoc for it. To have a typecheck add `implements SceneCollectionsServiceApi` here.

File: app/services/scene-collections/scene-collections-api.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+export interface ISceneCollectionsServiceApi {
+  load(id: string, shouldAttemptRecovery?: boolean): Promise<void>;
+  create(name?: string, setupFunction?: () => void): Promise<void>;
```
Comment (holiber) [inline app/services/scene-collections/scene-collections-api.ts]: We can't expose functions as arguments to the API. You can just omit the second param here.

File: app/services/scene-collections/scene-collections-api.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+export interface ISceneCollectionsServiceApi {
+  load(id: string, shouldAttemptRecovery?: boolean): Promise<void>;
```
Comment (holiber) [inline app/services/scene-collections/scene-collections-api.ts]: We need some jsdoc comments here for docs

File: app/services/scene-collections/scene-collections.ts [hunk 4]
```diff
@@ -0,0 +1,865 @@
...
+  activeId: string;
+  collections: ISceneCollectionsManifestEntry[];
+}
+
+/**
+ * V2 of the scene collections service:
+ * - Completely asynchronous
+ * - Server side backup
+ */
+export class SceneCollectionsService extends StatefulService<ISceneCollectionsManifest> {
```
Comment (holiber) [inline app/services/scene-collections/scene-collections.ts]: `implements ISceneCollectionsServiceApi`

File: app/services/scene-collections/scene-collections.ts [hunk 5]
```diff
@@ -0,0 +1,865 @@
...
+                      (sceneItemData: ISceneItemInfo) => {
+                        return {
+                          sceneItemId: sceneItemData.id,
+                          sourceId: sceneItemData.sourceId
+                        };
+                      }
+                    )
+                  };
+                }
+              ),
```
Comment (holiber) [inline app/services/scene-collections/scene-collections.ts]: omg, is it autoformatting? Looks like the Lisp hell

File: app/services/scene-collections/scene-collections.ts [hunk 6]
```diff
@@ -0,0 +1,865 @@
...
+    this.ADD_COLLECTION(id, name, (new Date()).toISOString());
+  }
+
+  /**
+   * Deletes on the server and removes from the store
+   */
+  private async removeCollection(id: string) {
+    this.DELETE_COLLECTION(id);
+
+    // TODO: Delete the file (for safety and recoverability we soft delete for now)
```
Comment (holiber) [inline app/services/scene-collections/scene-collections.ts]: Should  this TODO be a part of this PR?

File: app/services/scene-collections/scene-collections.ts [hunk 7]
```diff
@@ -0,0 +1,865 @@
...
+    } catch (e) {
+      console.error('Scene collection sync failed: ', e);
+    }
+  }
+
+  /**
+   * Duplicates the current scene collection
+   * @param name the name of the new scene collection
+   */
+  async duplicate(name: string) {
```
Comment (holiber) [inline app/services/scene-collections/scene-collections.ts]: it is not obvious that this method also switch the scene collection. Maybe rename it to something like `duplicateAndSetActive` ?

File: app/services/scene-collections/scene-collections.ts [hunk 8]
```diff
@@ -0,0 +1,865 @@
...
+  /**
+   * Install a new overlay from a URL
+   * @param url the URL of the overlay file
+   * @param name the name of the overlay
+   * @param progressCallback a callback that receives progress of the download
+   */
+  async installOverlay(url: string, name: string, progressCallback?: (info: IDownloadProgress) => void) {
+    this.startLoadingOperation();
+
+    // TODO: Handle Errors
```
Comment (holiber) [inline app/services/scene-collections/scene-collections.ts]: should this TODO be a part of this PR?

File: app/services/scene-collections/scene-collections-api.ts (hunk 2 above)
Comment (andycreeth) [reply app/services/scene-collections/scene-collections-api.ts]: 👍 It should be used over the API anyway.  It's just a hack for the OBS importer to use.

File: app/services/scene-collections/scene-collections.ts (hunk 8 above)
Comment (andycreeth) [reply app/services/scene-collections/scene-collections.ts]: Yeah, meant to do that.

File: app/services/scene-collections/scene-collections.ts (hunk 5 above)
Comment (andycreeth) [reply app/services/scene-collections/scene-collections.ts]: Haha yeah I guess so.  This function does have a lot of indentation, which the auto formatter seems to strange things with.

File: app/services/scene-collections/scene-collections.ts (hunk 6 above)
Comment (andycreeth) [reply app/services/scene-collections/scene-collections.ts]: Yeah I'm still undecided as to whether we should hard delete files.  I'll make a final decision and remove the comment.

**Comment by andycreeth** [comment]
Thanks for reviewing, I know it's a big PR.  It looks like this is going to be on hold for at least another day due to current memory leak issues on staging.  So I am going to take advantage of that time to write a bunch of tests.

**Comment by holiber** [comment]
We should update `docs/README.md` `How to switch a scene collection?` section

**Pull request merged**